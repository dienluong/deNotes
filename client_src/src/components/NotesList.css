/* NotesList styles */
.active-tree-node {
  box-shadow: 0 0 0 4px pink;
}
