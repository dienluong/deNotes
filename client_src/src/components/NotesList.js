/* TODO: Revisit this component and react-ui-tree code to see if rendering can be optimized
   https://reactjs.org/docs/optimizing-performance.html
*/
// import 'react-ui-tree/dist/react-ui-tree.css';
// import Tree from 'react-ui-tree';
// TODO: Delete Note.js
// import Note from './Note';

import Tree, { removeNode } from 'react-sortable-tree';
import 'react-sortable-tree/style.css';
// TODO: Delete NotesList.css
// import './NotesList.css';
import React from 'react';

import sampleNotes from '../test/sample-tree';

class NotesList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      notesTree: sampleNotes,
    };

    this.handleChange = this.handleChange.bind(this);
    // this.renderNode = this.renderNode.bind(this);
    // this.onClickNode = this.onClickNode.bind(this);
    this.deleteNote = this.deleteNote.bind(this);
  }

  handleChange(notesTree) {
    this.setState({
      notesTree,
    });
  }

  /* TODO: Remove, formerly for react-ui-tree
  renderNode(node) {
    // console.log(`renderNode ${node.module}`);
    return (
      <Note
        active={ this.state.activeNote === node }
        clickHandler={ this.onClickNode }
        node={ node }
      />
    );
  }

  onClickNode(node) {
    // console.log(`Active Note: ${node.module}`);
    this.setState({
      activeNote: node,
    });
  }
  */

  deleteNote({ path }) {
    return ({
      buttons: [
        <button
          style={{ verticalAlign: 'middle' }}
          onClick={() => {
            const { treeData } = removeNode({
              treeData: this.state.notesTree,
              getNodeKey: ({ treeIndex }) => treeIndex,
              path,
            });
            this.setState({
              notesTree: treeData,
            });
          }}
        >
          x
        </button>,
      ],
    });
  }

  render() {
    console.log('rendering NoteLists........................');
    return (
      <Tree
        className='tree'
        treeData={ this.state.notesTree }
        onChange={ this.handleChange }
        generateNodeProps={ this.deleteNote }
      />
    );
  };
}

export default NotesList;
