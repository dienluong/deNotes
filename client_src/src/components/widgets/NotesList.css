/* NotesList styles */
.dnt__active-tree-node {
  box-shadow: 0 0 0 4px pink;
}

.dnt__tree-node-btn {
  vertical-align: middle;
}
