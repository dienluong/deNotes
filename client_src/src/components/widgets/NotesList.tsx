import React from 'react';
import Toolbar from './Toolbar';
import PathNavigator from './PathNavigator';
import Tree from 'react-sortable-tree';
import mobileTheme from '../../tree-theme';
import 'react-sortable-tree/style.css';
import styles from './NotesList.module.css';
import NodeTitle from './NodeTitle';

// Types
import { TreeItem } from 'react-sortable-tree';
type PropsT = {
  tree: TreeNodeT[],
  activeNode: ActiveNodeT,
  activePath: string[],
  treeChangeHandler: (...args: any) => any;
  nodeTitleChangeHandler: (...args: any) => any,
  nodeClickHandler: (params: { id: TreeNodeT["id"], path: TreeNodePathT }) => unknown,
  nodeDoubleClickHandler: (params: { id: TreeNodeT["id"], path: TreeNodePathT }) => unknown,
  deleteNodeBtnHandler: (params: { node: TreeNodeT, path: TreeNodePathT }) => unknown,
  addNoteBtnHandler: (params: { path: TreeNodePathT }) => unknown,
  pathNavigatorClickHandler: (...args: any) => any,
  toolbarHandlersMap: any,
  getNodeKey: (...args: any) => any,
};
type generateNodePropsT = ({ node, path }: { node: TreeItem, path: (string|number)[] }) => object;

function NotesList({
  tree,
  activeNode,
  activePath,
  treeChangeHandler,
  nodeTitleChangeHandler,
  nodeClickHandler,
  nodeDoubleClickHandler,
  deleteNodeBtnHandler,
  addNoteBtnHandler,
  pathNavigatorClickHandler,
  toolbarHandlersMap,
  getNodeKey,
}: PropsT) {
  function buildNodeProps({ node, path }: { node: TreeNodeT, path: string[] }) {
    return ({
      title: (<NodeTitle node={ node } path={ path } onSubmit={ nodeTitleChangeHandler } />),
      className: (node.id === activeNode.id) ? `${styles['dnt__tree-node']} ${styles['dnt__tree-node--active']}` : styles['dnt__tree-node'],
      buttons: _buildNodeButtons({ node, path }),
      tabIndex: '0',
      'data-testid': node.id,
      onClick: () => nodeClickHandler({ id: node.id, path }),
      onDoubleClick: () => nodeDoubleClickHandler({ id: node.id, path }),
    });
  }

  function _buildNodeButtons({ node, path }: { node: TreeNodeT, path: string[] }) {
    let buttons = [
      <button
        className={ styles['dnt__tree-node-btn'] }
        onClick={ (event) => {
          event.stopPropagation();
          deleteNodeBtnHandler({ node, path });
        }}
      >
        x
      </button>,
    ];

    // Check if current node is parent node; '+' button only for parent nodes
    if (typeof node.children !== 'undefined') {
      buttons.unshift(
        <button
          className={ styles['dnt__tree-node-btn'] }
          onClick={ (event) => {
            event.stopPropagation();
            addNoteBtnHandler({ path });
          }}
        >
          +
        </button>);
    }

    return buttons;
  }

  function _collapseAllParents({ tree }: { tree: TreeNodeT[] }): TreeNodeT[] {
    return tree.map((node): TreeNodeT => {
      if (node && node.children) {
        return { ...node, expanded: false };
      } else {
        return node;
      }
    });
  }

  return (
    <div className={ styles["dnt__notes-list"] }>
      <Toolbar toolsMap={ toolbarHandlersMap } />
      <PathNavigator
        path={ activePath }
        activeSegmentIdx={ activeNode.path.indexOf(activeNode.id) }
        onClick={ pathNavigatorClickHandler }
      />
      <Tree
        className={ 'tree ' + styles.dnt__tree }
        treeData={ _collapseAllParents({ tree }) }
        theme={ mobileTheme as any }
        onChange={ treeChangeHandler }
        getNodeKey={ getNodeKey }
        generateNodeProps={ buildNodeProps as generateNodePropsT }
      />
    </div>
  );
}

export default NotesList;
