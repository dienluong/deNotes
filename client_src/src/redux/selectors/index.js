import { find } from 'react-sortable-tree';
import { createSelector } from 'reselect';
import { getNodeKey } from '../../utils/treeUtils';
import { selectNotesTreeTree, selectActiveNodePath } from '../reducers';

// TODO: remove
// const selectNotesTree = (state) => state.notesTree;
// const selectActiveNodePath = (state) => state.activeNode.path;

/**
 * For each note ID in path, return the corresponding title.
 * @param notesTree {!Object[]}
 * @param idList {Array}
 * @return {Array}
 * @private
 */
export const selectTitlesFromActivePath = createSelector(
  [selectNotesTreeTree, selectActiveNodePath],
  (notesTree = [], idList = []) => {
    if (!Array.isArray(idList) || !idList.length) {
      return [];
    }

    return idList.map((id) => {
      const matches = find({
        getNodeKey,
        treeData: notesTree,
        searchQuery: id,
        searchMethod: ({ node, searchQuery }) => searchQuery === node.id,
      }).matches;
      return matches.length ? matches[0].node.title : '';
    });
  }
);
