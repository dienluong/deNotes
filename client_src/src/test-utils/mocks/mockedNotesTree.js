import { nodeTypes } from '../../utils/appCONSTANTS';

const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

export const mockedTree = [
  {
    title: 'folder root[0]',
    subtitle: '',
    uniqid: '1',
    type: nodeTypes.FOLDER,
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    expanded: false,
    children: [
      {
        title: 'note root[0].children[0]',
        subtitle: '',
        uniqid: '2',
        type: nodeTypes.ITEM,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
      },
      {
        title: 'note root[0].children[1]',
        subtitle: '',
        uniqid: '3',
        type: nodeTypes.ITEM,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
      },
      {
        title: 'folder root[0].children[2]',
        subtitle: '',
        uniqid: '4',
        type: nodeTypes.FOLDER,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        expanded: false,
        children: [
          {
            title: 'note root[0].children[2].children[0]',
            subtitle: '',
            uniqid: '5',
            type: nodeTypes.ITEM,
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
          },
          {
            title: 'note root[0].children[2].children[1]',
            subtitle: '',
            uniqid: '6',
            type: nodeTypes.ITEM,
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
          },
        ],
      },
      {
        title: 'folder root[0].children[3]',
        subtitle: '',
        uniqid: '7',
        type: nodeTypes.FOLDER,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        expanded: false,
        children: [],
      },
    ],
  },
  {
    title: 'note root[1]',
    subtitle: '',
    uniqid: '8',
    type: nodeTypes.ITEM,
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
  },
  {
    title: 'folder root[2]',
    subtitle: '',
    uniqid: '9',
    type: nodeTypes.FOLDER,
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    expanded: false,
    children: [],
  },
];
