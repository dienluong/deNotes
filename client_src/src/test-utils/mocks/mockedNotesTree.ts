import { nodeTypes } from '../../utils/appCONSTANTS';

const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

export const mockedTree: TreeNodeT[] = [
  {
    title: 'folder root0',
    subtitle: '',
    uniqid: '1',
    type: nodeTypes.FOLDER,
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    expanded: true,
    children: [
      {
        title: 'note root0.children0',
        subtitle: '',
        uniqid: '2',
        type: nodeTypes.ITEM,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
      },
      {
        title: 'note root0.children1',
        subtitle: '',
        uniqid: '3',
        type: nodeTypes.ITEM,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
      },
      {
        title: 'folder root0.children2',
        subtitle: '',
        uniqid: '4',
        type: nodeTypes.FOLDER,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        expanded: true,
        children: [
          {
            title: 'note root0.children2.children0',
            subtitle: '',
            uniqid: '5',
            type: nodeTypes.ITEM,
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
          },
          {
            title: 'note root0.children2.children1',
            subtitle: '',
            uniqid: '6',
            type: nodeTypes.ITEM,
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
          },
        ],
      },
      {
        title: 'folder root0.children3',
        subtitle: '',
        uniqid: '7',
        type: nodeTypes.FOLDER,
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        expanded: true,
        children: [],
      },
    ],
  },
  {
    title: 'note root1',
    subtitle: '',
    uniqid: '8',
    type: nodeTypes.ITEM,
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
  },
  {
    title: 'folder root2',
    subtitle: '',
    uniqid: '9',
    type: nodeTypes.FOLDER,
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    expanded: true,
    children: [],
  },
];
