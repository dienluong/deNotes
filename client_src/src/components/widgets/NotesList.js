import React, { Fragment } from 'react';
import Toolbar from './Toolbar';
import PathNavigator from './PathNavigator';
import Tree from 'react-sortable-tree';
import 'react-sortable-tree/style.css';
import './NotesList.css';
import NodeTitle from './NodeTitle';

function NotesList({
  notesTree,
  activeNode,
  activePath,
  treeChangeHandler,
  nodeTitleChangeHandler,
  nodeClickHandler,
  deleteNodeBtnHandler,
  addNoteBtnHandler,
  pathNavigatorClickHandler,
  toolbarHandlersMap,
  getNodeKey,
}) {
  function buildNodeProps({ node, path }) {
    return ({
      title: (<NodeTitle node={ node } path={ path } onSubmit={ nodeTitleChangeHandler } />),
      className: (node.id === activeNode.id) ? 'dnt__active-tree-node' : '',
      buttons: _buildNodeButtons({ node, path }),
      tabIndex: '0',
      onClick: () => nodeClickHandler({ id: node.id, path }),
      onFocus: () => console.log('Focused: ' + node.title),
      onBlur: () => console.log('unFocused: ' + node.title),
    });
  }

  function _buildNodeButtons({ node, path }) {
    let buttons = [
      <button
        className='dnt__tree-node-btn'
        onClick={ (event) => {
          event.stopPropagation();
          deleteNodeBtnHandler({ node, path });
        }}
      >
        x
      </button>,
    ];

    // Check if current node is parent node; '+' button only for parent nodes
    if (typeof node.children !== 'undefined') {
      buttons.unshift(
        <button
          className='dnt__tree-node-btn'
          onClick={ (event) => {
            event.stopPropagation();
            addNoteBtnHandler({ path });
          }}
        >
          +
        </button>);
    }

    return buttons;
  }

  // TODO: remove
  // console.log(`
  //   Tree: ${JSON.stringify(notesTree)} \n
  //   Active ID: ${activeNode.id} \n
  //   Path: ${activeNode.path} \n
  //   ${ activePath }
  // `);

  return (
    <Fragment>
      <Toolbar toolsMap={ toolbarHandlersMap } />
      <PathNavigator
        path={ activePath }
        activeSegmentIdx={ activeNode.path.indexOf(activeNode.id) }
        onClick={ pathNavigatorClickHandler }
      />
      <Tree
        className='tree'
        treeData={ notesTree }
        onChange={ treeChangeHandler }
        getNodeKey={ getNodeKey }
        generateNodeProps={ buildNodeProps }
      />
    </Fragment>
  );
}

export default NotesList;
