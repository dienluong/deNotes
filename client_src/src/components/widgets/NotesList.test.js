import React from 'react';
import NotesList from './NotesList';
import notesListStyles from './NotesList.module.css';
import pathNavStyles from './PathNavigator.module.css';
import { render, cleanup, fireEvent, within, getNodeText } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { walk, find, getFlatDataFromTree } from 'react-sortable-tree';
import { getNodeKey } from '../../utils/treeUtils';
import baseState from '../../redux/misc/initialState';
import { selectTitlesFromActivePath } from '../../redux/selectors';
import { mockedTree } from '../../test-utils/mocks/mockedNotesTree';

afterEach(cleanup);

it('should render all node titles in the first level of the received tree', () => {
  const tree = mockedTree;
  const activeNode = baseState.activeNode;
  const activePath = selectTitlesFromActivePath(baseState);
  const toolbarHandlersMap = new Map();
  toolbarHandlersMap.set('tool1', jest.fn());
  toolbarHandlersMap.set('tool2', jest.fn());
  const props = {
    tree,
    activeNode,
    activePath,
    treeChangeHandler: jest.fn(),
    nodeTitleChangeHandler: jest.fn(),
    nodeClickHandler: jest.fn(),
    deleteNodeBtnHandler: jest.fn(),
    addNoteBtnHandler: jest.fn(),
    pathNavigatorClickHandler: jest.fn(),
    toolbarHandlersMap,
    getNodeKey,
  };

  const { container, queryAllByValue } = render(<NotesList { ...props }/>);

  expect(container).toMatchSnapshot();
  const renderedInputs = Array.from(container.getElementsByTagName('input'));
  const renderedTitles = renderedInputs.map(input => input.value);
  // const renderedTitles = renderedInputs.map(getNodeText);
  console.log(renderedTitles);

  const test = (node) => {
    expect(queryAllByValue(node.title)).toHaveLength(1);
    expect(renderedTitles.includes(node.title)).toBe(true);
  };

  props.tree.forEach(test);

  // walk({
  //   treeData: props.tree,
  //   getNodeKey,
  //   callback: test,
  //   ignoreCollapsed: false,
  // });

  const numOfRenderedTreeNodeElements = container.getElementsByClassName(notesListStyles['dnt__tree-node']).length;
  // const numOfNodesInSourceTree = getFlatDataFromTree({ treeData: props.tree, getNodeKey, ignoreCollapsed: false }).length;
  const numOfNodesInSourceTree = props.tree.length;
  expect(numOfRenderedTreeNodeElements).toEqual(numOfNodesInSourceTree);
});

it('should invoke handler when tree node is clicked', () => {
  const tree = mockedTree;
  const activeNode = baseState.activeNode;
  const activePath = selectTitlesFromActivePath(baseState);
  const toolbarHandlersMap = new Map();
  toolbarHandlersMap.set('tool1', jest.fn());
  toolbarHandlersMap.set('tool2', jest.fn());
  const props = {
    tree,
    activeNode,
    activePath,
    treeChangeHandler: jest.fn(),
    nodeTitleChangeHandler: jest.fn(),
    nodeClickHandler: jest.fn(),
    deleteNodeBtnHandler: jest.fn(),
    addNoteBtnHandler: jest.fn(),
    pathNavigatorClickHandler: jest.fn(),
    toolbarHandlersMap,
    getNodeKey,
  };

  const { container } = render(<NotesList { ...props }/>);

  const renderedTreeNodeElements = [...container.getElementsByClassName(notesListStyles['dnt__tree-node'])];
  renderedTreeNodeElements.forEach(element => {
    const elementTitle = element.getElementsByTagName('input')[0].value;
    const correspondingNode = find({
      treeData: props.tree,
      getNodeKey,
      searchQuery: elementTitle,
      searchMethod: ({ node, searchQuery }) => node.title === searchQuery,
    }).matches;
    // This assertion is NOT the actual test. For each rendered element, we should have one and only one corresponding node in the treeData.
    expect(correspondingNode).toHaveLength(1);

    fireEvent.click(element);
    // this assertion is the actual test
    expect(props.nodeClickHandler).lastCalledWith({ id: correspondingNode[0].node.id, path: correspondingNode[0].path });
  });

  const numberOfNodes = getFlatDataFromTree({ treeData: props.tree, getNodeKey, ignoreCollapsed: false }).length;
  expect(props.nodeClickHandler).toBeCalledTimes(numberOfNodes);
});

it('should invoke handlers when toolbar button is clicked', () => {
  const tree = mockedTree;
  const activeNode = baseState.activeNode;
  const activePath = selectTitlesFromActivePath(baseState);
  const toolbarHandlersMap = new Map();
  toolbarHandlersMap.set('tool1', jest.fn());
  toolbarHandlersMap.set('tool2', jest.fn());
  toolbarHandlersMap.set('tool3', jest.fn());
  const props = {
    tree,
    activeNode,
    activePath,
    treeChangeHandler: jest.fn(),
    nodeTitleChangeHandler: jest.fn(),
    nodeClickHandler: jest.fn(),
    deleteNodeBtnHandler: jest.fn(),
    addNoteBtnHandler: jest.fn(),
    pathNavigatorClickHandler: jest.fn(),
    toolbarHandlersMap,
    getNodeKey,
  };

  const { queryAllByText } = render(<NotesList { ...props }/>);

  for (let [label, handler] of toolbarHandlersMap) {
    const button = queryAllByText(label);
    expect(button).toHaveLength(1);
    fireEvent.click(button[0]);
    expect(handler).toBeCalled();
  }

  toolbarHandlersMap.forEach(handler => expect(handler).toBeCalledTimes(1));
});

it('should invoke handler when segment in path navigator is clicked', () => {
  const tree = mockedTree;
  const activeNode = baseState.activeNode;
  const activePath = selectTitlesFromActivePath(baseState);
  const toolbarHandlersMap = new Map();
  toolbarHandlersMap.set('tool1', jest.fn());
  toolbarHandlersMap.set('tool2', jest.fn());
  const props = {
    tree,
    activeNode,
    activePath,
    treeChangeHandler: jest.fn(),
    nodeTitleChangeHandler: jest.fn(),
    nodeClickHandler: jest.fn(),
    deleteNodeBtnHandler: jest.fn(),
    addNoteBtnHandler: jest.fn(),
    pathNavigatorClickHandler: jest.fn(),
    toolbarHandlersMap,
    getNodeKey,
  };

  const { container } = render(<NotesList { ...props }/>);
  expect(container.getElementsByClassName(pathNavStyles.dnt__pathnav)).toHaveLength(1);
  const renderedPathNav = container.getElementsByClassName(pathNavStyles.dnt__pathnav).item(0);
  activePath.forEach((segment, idx) => {
    const renderedPathSegment = within(renderedPathNav).getByText(new RegExp(`${segment}`));
    fireEvent.click(renderedPathSegment);
    expect(props.pathNavigatorClickHandler).lastCalledWith({ idx });
  });
});

it('should invoke handler when tree node button is clicked', () => {
  const tree = mockedTree;
  const activeNode = baseState.activeNode;
  const activePath = selectTitlesFromActivePath(baseState);
  const toolbarHandlersMap = new Map();
  toolbarHandlersMap.set('tool1', jest.fn());
  toolbarHandlersMap.set('tool2', jest.fn());
  const props = {
    tree,
    activeNode,
    activePath,
    treeChangeHandler: jest.fn(),
    nodeTitleChangeHandler: jest.fn(),
    nodeClickHandler: jest.fn(),
    deleteNodeBtnHandler: jest.fn(),
    addNoteBtnHandler: jest.fn(),
    pathNavigatorClickHandler: jest.fn(),
    toolbarHandlersMap,
    getNodeKey,
  };

  const { getByTestId } = render(<NotesList { ...props }/>);

  let numberOfFolders = 0;
  const test = ({ node, path }) => {
    if (node.type === 'folder') {
      numberOfFolders += 1;
      const renderedTreeNode = getByTestId(node.id);
      const renderedTitles = within(renderedTreeNode).queryAllByValue(node.title);
      expect(renderedTitles).toHaveLength(1);
      const nodeAddButton = within(renderedTreeNode).getByText('+');
      const nodeDeleteButton = within(renderedTreeNode).getByText('x');

      fireEvent.click(nodeAddButton);
      expect(props.addNoteBtnHandler).lastCalledWith({ path });
      expect(props.addNoteBtnHandler).toBeCalledTimes(numberOfFolders);

      fireEvent.click(nodeDeleteButton);
      expect(props.deleteNodeBtnHandler).lastCalledWith({ node, path });
      expect(props.deleteNodeBtnHandler).toBeCalledTimes(numberOfFolders);
    }
  };

  walk({
    treeData: props.tree,
    getNodeKey,
    callback: test,
    ignoreCollapsed: false,
  });
});

it('should invoke handler on node title change', () => {
  const tree = mockedTree;
  const activeNode = baseState.activeNode;
  const activePath = selectTitlesFromActivePath(baseState);
  const toolbarHandlersMap = new Map();
  toolbarHandlersMap.set('tool1', jest.fn());
  toolbarHandlersMap.set('tool2', jest.fn());
  const props = {
    tree,
    activeNode,
    activePath,
    treeChangeHandler: jest.fn(),
    nodeTitleChangeHandler: jest.fn(),
    nodeClickHandler: jest.fn(),
    deleteNodeBtnHandler: jest.fn(),
    addNoteBtnHandler: jest.fn(),
    pathNavigatorClickHandler: jest.fn(),
    toolbarHandlersMap,
    getNodeKey,
  };

  const { queryAllByValue } = render(<NotesList { ...props }/>);

  let numberOfNodes = 0;
  const test = ({ node, path }) => {
    const newTitle = 'new test title';
    const renderedInputs = queryAllByValue(node.title);
    expect(renderedInputs).toHaveLength(1);
    const renderedTitle = renderedInputs[0];

    numberOfNodes += 1;
    renderedTitle.value = newTitle;
    fireEvent.blur(renderedTitle);
    expect(props.nodeTitleChangeHandler).lastCalledWith({ node, path, title: newTitle });
    expect(props.nodeTitleChangeHandler).toBeCalledTimes(numberOfNodes);
  };

  walk({
    treeData: props.tree,
    getNodeKey,
    callback: test,
    ignoreCollapsed: false,
  });
});
