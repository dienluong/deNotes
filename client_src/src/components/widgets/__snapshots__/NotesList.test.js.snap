// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render all node titles in the first level of the received tree and highlight the active node 1`] = `
<div>
  <div
    class="dnt__notes-list"
  >
    <div
      class="dnt__toolbar"
    >
      <button
        class="dnt__tool"
      >
        tool1
      </button>
      <button
        class="dnt__tool"
      >
        tool2
      </button>
    </div>
    <nav
      class="dnt__pathnav"
    >
      <ul>
        <li
          class="lined dnt__pathnav-segment--active"
        >
          <span>
            root-folder 1
            &gt;
          </span>
        </li>
      </ul>
    </nav>
    <div
      class="rst__tree tree dnt__tree"
      style="height: 100%;"
    >
      <div>
        <div
          aria-label="grid"
          aria-readonly="true"
          class="ReactVirtualized__Grid ReactVirtualized__List rst__virtualScrollOverride"
          role="grid"
          style="box-sizing: border-box; direction: ltr; height: 99999px; position: relative; width: 200px; overflow-x: hidden; overflow-y: hidden;"
          tabindex="0"
        >
          <div
            class="ReactVirtualized__Grid__innerScrollContainer"
            role="rowgroup"
            style="width: auto; height: 186px; max-width: 200px; max-height: 186px; overflow: hidden; position: relative;"
          >
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 0px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="FOLDER_NODE|^|1"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div>
                    <button
                      aria-label="Collapse"
                      class="expandButton"
                      type="button"
                    />
                  </div>
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node dnt__tree-node--active"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="root-folder 1"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              +
                            </button>
                          </div>
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 62px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="ITEM_NODE|^|8"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="root-note 1"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 124px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="FOLDER_NODE|^|9"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="root-folder 2"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              +
                            </button>
                          </div>
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
